[새기능] 캐릭터 선택기능 구현

jyj1289 opened the pull request:
## 🎫 관련 이슈
[//]: # (다음 키워드를 사용하면 해당 PR을 머지할 때 자동으로 이슈를 닫을 수 있습니다.)
[//]: # (keyword: close|closes|closed|resolve|resolves|resolved|fix|fixes|fixed)
[//]: # (예시: close #1)

close #4

<br>

## 📄 개요
[//]: # (작업 내용을 간단히 요약해서 적습니다.)
[//]: # (예시: 유저 회원가입 기능을 만들었습니다.)

> 캐릭터 선택 기능을 만들었습니다.

<br>

## 🔨 작업 내용
[//]: # (작업 내용을 자세하게 적습니다.)
[//]: # (붙임표 "-" 을 사용해서 목록을 만듭니다.)
[//]: # (예시: 유저 회원가입 API를 만들었습니다.)

- 캐릭터 엔티티를 구현했습니다.
- 캐릭터 Facade를 구현했습니다.
- 캐릭터 Controller를 구현했습니다.
- 캐릭터 선택 API를 만들었습니다.



<br>

## 🏁 확인 사항
[//]: # (PR을 보내기 전 다음 사항을 확인해주세요.)
[//]: # (해당 사항을 모두 이행해야 머지할 수 있습니다.)
[//]: # (- [x] 를 사용해서 완료로 표시할 수 있습니다.)

- [x] 테스트를 완료했나요?
- [x] API 문서를 작성했나요?
- [x] 코드 컨벤션을 준수했나요?
- [x] 불필요한 로그, 주석, import 등을 삭제했나요?

<br>

## 🙋🏻 덧붙일 말
[//]: # (다음 사항이 있다면 적어주세요.)
[//]: # (PR에 대한 추가 설명)
[//]: # (중점적으로 리뷰받고 싶은 부분)
[//]: # (기타 등등)

jyj1289 committed:
feat(#4): 캐릭터 선택 API 구현
- 캐릭터 엔티티 및 레포지토리를 구현했어요.
- 캐릭터 선택 API를 구현했어요.
- CharacterType을 정의했어요.
- CharacterFacade를 구현했어요.

jyj1289 committed:
test(#4): 캐릭터 선택 API 테스트코드 작성
- docs 작성을 위한 캐릭터 선택 API 컨트롤러 테스트 코드를 작성했어요.

jyj1289 committed:
docs(#4): 캐릭터 선택 API 문서화
- 캐릭터 선택 API를 문서화했어요.

jyj1289 committed:
refactor(#4): NoArgsConstructor 접근 제어자 변경
- NoArgsConstructor 접근 제어자를 PROTECTED로 변경했어요.

cabbage16 committed:
Merge branch 'develop' into feat/#4

cabbage16 committed:
fix(#4): EnumProperty 경로 변경에 따른 수정
- develop 브랜치에서 EnumProperty의 경로가 변경되어서 적용했어요.

jyj1289 committed:
Merge remote-tracking branch 'origin/feat/#4' into feat/#4

jyj1289 committed:
feat(#4): Exception 핸들링 및 문서화
- 개별 exception을 만들었어요.
- 정상적이지 않은 응답을 문서화했어요.

jyj1289 committed:
refactor(#4): 불필요한 어노테이션 삭제
- Transactional 어노테이션을 붙이지 않아도 작동하기 때문에 제거했어요.

gimhanul commented on src/main/java/com/bamdoliro/sinabro/application/character/SelectCharacterUseCase.java:
```diff
@@ -0,0 +1,37 @@
+    public void execute(User user, SelectCharacterRequest request) {
+        validate(user);
+
+        Character character = Character.builder()
+                .type(request.getType())
+                .user(user)
+                .build();
```
builder 사용은 지양하는 게 좋다고 생각합니당
빠뜨리는 parameter가 생길 수 있어서요

gimhanul commented on src/main/java/com/bamdoliro/sinabro/application/character/SelectCharacterUseCase.java:
```diff
@@ -0,0 +1,37 @@
+                .user(user)
+                .build();
+
+        characterRepository.save(character);
```
`@Transactional` annotaion 누락된 거 같아요~

cabbage16 replied:
JpaRepository의 save 메서드를 사용하면 `@Transactional`을 사용하지 않아도 상관없지 않나요?

gimhanul commented on src/main/java/com/bamdoliro/sinabro/application/character/SelectCharacterUseCase.java:
```diff
@@ -0,0 +1,37 @@
+
+    private void validate(User user) {
+        Optional<Character> character = characterRepository.findByUser(user);
+        if (character.isPresent()) {
```
요것도 functional하게 .ifPresent() 로 할 수 있을 거 같은데 아닐 수도 있을듯 확인해주세요!

gimhanul commented on src/main/java/com/bamdoliro/sinabro/domain/character/domain/Character.java:
```diff
@@ -0,0 +1,40 @@
+    @Column(nullable = false)
+    private Integer friendship;
+
+    @OneToOne(fetch = FetchType.LAZY)
```
OneToOne은 column을 가지고 있는 주체가 누군지 중요한데 충분히 고려해주세요 ㅎ
회사에서 몰래 리뷰 중이라 깊게 생각할 여력이 없네용 ㅎ

cabbage16 commented on src/main/java/com/bamdoliro/sinabro/presentation/character/dto/request/SelectCharacterRequest.java:
```diff
@@ -0,0 +1,17 @@
+import lombok.NoArgsConstructor;
+
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
```
P3) RequestDto에는 왜 굳이 AccessLevel을 Protected로 한 것인지 궁금해요

jyj1289 committed:
refactor(#4): Transactional 어노테이션 추가
- 불필요한 줄 알았던 Transactional 어노테이션이 커밋과 롤백을 해준다는 것을 알고 부활시켰어요.

jyj1289 committed:
refactor(#4): Builder 어노테이션 삭제 및 validate 메서드 리팩토링
- 파라미터가 적어 Builder를 사용하지 않는 것이 더욱 안정적일것이라고 생각했어요.
- 캐릭터 중복 검증을 할 때 findByUser를 호출하여 isPresent()로 하는 것보다 existsByUser를 호출하는 것이 더 효율적이라 판단했어요.

jyj1289 committed:
refactor(#4): Builder 어노테이션 삭제 및 validate 메서드 리팩토링
- 파라미터가 적어 Builder를 사용하지 않는 것이 더욱 안정적일것이라고 생각했어요.
- 캐릭터 중복 검증을 할 때 findByUser를 호출하여 isPresent()로 하는 것보다 existsByUser를 호출하는 것이 더 효율적이라 판단했어요.
